### ISSUE TITLE
Proteins management in GNB

### ISSUE THREAD

**Issue by mboudet** [holmrenser/bioportal#304]
Hello,

While working on the orthogroups PR, I ran into an issue with the way proteins are managed in GNB.

Currently, there are no 'proteins' in DB. When accessing a gene, proteins sequences are translated on the fly based on transcripts.
This leads to some issues:

1) Proteins are translated every time someone access a gene page. 
(Well, to be fair, the translation is not the most computation-heavy part of rendering. I feel like everything in [getGeneSequences](https://github.com/genenotebook/genenotebook/blob/master/imports/api/util/util.js#L168) should be done once in the back, and not every time at render)

2) The protein id will be the same as the mRNA. In our case, we usually add a suffix (-P) to our proteins to avoid mixup in software.
This means that files generated by functional annotations will contain these proteins names, and they will not be importable in GNB (as they will not match)

What could be done:

1) Store the proteins (and their sequences) in DB after loading a GFF. (Maybe as a subfeature?)
2) When loading GFFs, provide optional CLI options to generate the protein ID, such as:
  * A regex (to extract and generate from the mRNA ID)
  * A key, to be found in the GFF 'attributes' column
  * Nothing, which will defaut to the mRNA ID

3) It might be good to store the result of getGeneSequence in DB too, at the gene level, to avoid having to loop over transcripts & CDS every time we render a Gene page.

**Comment by holmrenser** [comment]
Interesting point, some things to consider:

- Computing on the fly means `genome + annotation = protein sequence` is always correct. If the protein sequence is stored in the DB we might need a procedure to verify this is still the case. Perhaps not very relevant now, but I wouldn't mind having functionality to change gene models in genenotebook at some point (similar to jbrowse/apollo).
- In my experience having identifiers that do not match is a bigger issue than mixing up mRNA with amino acid sequences. I'd be interested to see a use case where mRNA and amino acid sequences are easily confused.
- Would it be possible to get some numbers on the computational aspect of translating on the fly? My instinct says it is not a performance issue but I cannot back this up with data.

**Comment by mboudet** [comment]
Hello!

I think the computing cost for the sequences should be minor, as long as we only do it on a specific gene page (not on the gene list for example), or when downloading files.

I'm not sure I understand what you mean by mixing up "mRNA with amino acid sequences"? Our issue in our case is purely with the mRNA/proteins IDs. For instance, we currently have a Tripal instance, and having the same ID for both mRNA and proteins caused issues in DB, so we had to add a '-P' suffix to our proteins. 
So, we currently have 50+ genomes with a functional analysis done using 'suffixed' proteins, that cannot be integrated in GNB.

Leaving the computing sequences aside, we would need to store at least a list of protein IDs linked to a gene (without storing a sequence). By default, this list would be the same as the mRNA IDs, but we could pass options (regex, others) when loading a GFF to customize these proteins IDs.

**Comment by abretaud** [comment]
For the mrna/protein ids, I can add that it's very common to have different ids for them in data available in other databanks, like ensembl for example. We often need to load data from these databanks and can't change the ids

**Comment by holmrenser** [comment]
Thanks for clarifying and additional discussion! I think I can now see the use of having separate protein identifiers. Next is coming up with a clever/useful naming/linking scheme. I'm currently mostly leaning towards the way ensembl structures it's gff files: mRNA features with `biotype=protein_coding` seem to have CDS children that have a `protein_id` attribute. I don't think it's the most efficient way of doing this (I would think the mRNA feature has the protein_id attribute), but it nicely links transcript IDs and protein IDs. 

I think it should be straightforward to add an additional protein_id field to the genenotebook schema and use that where relevant.

**Comment by mboudet** [comment]
Agreed, though since many times gffs are not properly formated (especially in the attribute column), it would be best to provide several ways of generating these proteins ids, which would be selected by the user when loading data.

Currently our loading in Tripal works as follows:

* If nothing is provided, use the mRNA id and add '-protein'
* If the user provide a `protein_id_attr` key, we look to the mRNA/CDS attribute to get the id (ie, your `protein id` key)
* If the user provide a regex, (and an optional capture regex for the mRNA id), we use that to generate the protein ID.

Keeping the three options would allow GNB to be flexible in regard to GFF formating.

As for where to store the protein ID, we could maybe store it as subfeature, linked to the mRNA? Not quite sure yet, need to look at the loaders for the functional annotation part to find where it could fit, and allow for a simple DB query to get the related gene.
